### ISSUE TITLE
I never expected "foreach (var obj in objectsInBucket)" to throw an exception.

### ISSUE THREAD

**Issue by SurferJeffAtGoogle** [googleapis/google-cloud-dotnet#2694]
#### Environment details

  - OS:

Windows.

  - .NET version:

All.
  
- Package name and version:
```
    <PackageReference Include="Google.Cloud.Storage.V1" Version="2.3.0-beta02" />
```
#### Steps to reproduce

Full code here:
https://github.com/GoogleCloudPlatform/dotnet-docs-samples/blob/5d0fd9bab0d23a801c0dd3546ba815a865e80c5e/testutil/TestBucket.cs

I never expected the foreach statement in the code below to throw an exception.

```
                PagedEnumerable<Google.Apis.Storage.v1.Data.Objects,
                    Google.Apis.Storage.v1.Data.Object> objectsInBucket;
                try
                {
                    objectsInBucket = _storage.ListObjects(BucketName);
                }
                catch (Google.GoogleApiException e)
                when (e.Error.Code == 404)
                {
                    return;  // Bucket does not exist.  Ok.
                }

                // Try to delete each object in the bucket.
                foreach (var obj in objectsInBucket)
```

If there are no objects in the bucket, I expect the iterator to be empty.  If the bucket does not exist, I expect the call to ListObjects() to throw an exception.

**Comment by SurferJeffAtGoogle** [comment]
Here's the exception:
```
[xUnit.net 00:01:00.2953631]       Google.GoogleApiException : Google.Apis.Requests.RequestError
[xUnit.net 00:01:00.2955119]       Not Found [404]
[xUnit.net 00:01:00.2955827]       Errors [
[xUnit.net 00:01:00.2956416]       	Message[Not Found] Location[ - ] Reason[notFound] Domain[global]
[xUnit.net 00:01:00.2959794]       ]
[xUnit.net 00:01:00.2960785]       
```

**Comment by jskeet** [comment]
Hmm. I do have sympathy for this, but making the operation "fully lazy" makes various other aspects simpler and more consistent. It's also consistent with other lazy loading scenarios, e.g. LINQ against a database, where constructing the `IQueryable<T>` won't fail if the table has been deleted. I do understand that's a less likely scenario than a bucket not being present. (Note that this isn't just for storage - *all* our APIs that use pagination will behave the same way.)

Additionally, we can't change the design now even if after careful consideration we decided we wanted to, as moving when the exception is thrown would be a breaking change.

I think the best we can do is document this very clearly. In your experience, what do you think are the most important places to make this obvious? We should at least talk about it in https://googleapis.github.io/google-cloud-dotnet/docs/guides/page-streaming.html, but I suspect most users won't see that anyway.

**Comment by SurferJeffAtGoogle** [comment]
I put my cursor on `ListObjects`, press **F12**, and see this:

```cs
        //
        // Summary:
        //     Lists the objects in a given bucket, synchronously but lazily.
        //
        // Parameters:
        //   bucket:
        //     The bucket to list the objects from. Must not be null.
        //
        //   prefix:
        //     The prefix to match. Only objects with names that start with this string will
        //     be returned. This parameter may be null or empty, in which case no filtering
        //     is performed.
        //
        //   options:
        //     The options for the operation. May be null, in which case defaults will be supplied.
        //
        // Returns:
        //     A sequence of objects in the specified bucket.
        //
        // Remarks:
        //     This method fetches the objects lazily, making requests to the underlying service
        //     for a page of results at a time, as required.
        public virtual PagedEnumerable<Objects, Apis.Storage.v1.Data.Object> ListObjects(string bucket, string prefix = null, ListObjectsOptions options = null);
```

Could we add a note in the Returns on Remarks section that the iterating over the enumerator may throw an exception?

**Comment by jskeet** [comment]
This was fixed in #2753, although we haven't released the new Storage/BigQuery libraries yet.
